### PR TITLE
feat: add yeo-johnson step

### LINKED ISSUE #163: Yeo-Johnson

### PR DESCRIPTION
**Pull request by dshemetov**
close #163 

- the lambda parameter fitting is done via 0th order optimization i ripped from [recipes::yeo_johnson](https://recipes.tidymodels.org/reference/step_YeoJohnson.html)

TODO

- [x] layer skeleton in place
- [x] write inverse yj transform
- [x] tests
- [x] slather.layer_epi_YeoJohnson edge cases (see the TODO comment in the function... I'm not clear on what to expect from the `terms` interface in a layer)
- [x] get feedback
- [x] PR into epipredict cmu-delphi/epipredict#451

### CODE REVIEW

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+    skip = skip,
+    id = id
+  )
+}
+
+#' @export
+prep.step_epi_YeoJohnson <- function(x, training, info = NULL, ...) {
+  # Check that the columns selected for transformation are numeric.
+  col_names <- recipes_eval_select(x$terms, training, info)
+  check_type(training[, col_names], types = c("double", "integer"))
```
Comment (dajmcdon) [inline R/new_epipredict_steps/step_yeo_johnson.R:163]: Where is this function? In `recipes`?
```suggestion
  recipes::check_type(training[, col_names], types = c("double", "integer"))
```

**Review by dajmcdon — COMMENTED** [review]
I made it partway. It's looking good so far!

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+  recipe,
+  ...,
+  role = NA,
+  trained = FALSE,
+  lambdas = NULL,
+  na_lambda_fill = 1 / 4,
+  limits = c(-5, 5),
+  num_unique = 5,
+  na_rm = TRUE,
+  epi_keys_checked = NULL,
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply R/new_epipredict_steps/step_yeo_johnson.R:91]: They're exposed in the recipes step this is based on. Not sure why that exposes λ if you're not supposed to set it https://recipes.tidymodels.org/reference/step_YeoJohnson.html

[the remark this answers is not in the document]
Comment (dshemetov) [reply R/new_epipredict_steps/step_yeo_johnson.R:91]: (I think David's comment was meant for the lambdas = NULL comment above?

But yea, 
```
lambdas = NULL,
  limits = [c](https://rdrr.io/r/base/c.html)(-5, 5),
  num_unique = 5,
  na_rm = TRUE,
```
are copied from the original step. I don't seriously expect users to want to overwrite with their own lambdas though, so happy to remove that, if we don't care to keep the interface the same.)

`epi_keys_checked` ... I just lazily copied from step_adjust_ahead. I don't really anticipate grouping by anything other than geo_value + other_keys, so we can probably remove it and use x$template and objec$template below.

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+  keys <- key_colnames(new_data, exclude = "time_value")
+  if (!identical(keys, object$epi_keys_checked)) {
+    cli::cli_abort(
+      "The keys of the new data do not match the keys of the training data.",
+      call = rlang::caller_fn()
+    )
+  }
+  # Check that the columns for transformation are present in new_data.
+  col_names <- object$columns
+  check_new_data(col_names, object, new_data)
```
[the remark this answers is not in the document]
Comment (dshemetov) [reply R/new_epipredict_steps/step_yeo_johnson.R:217]: I only need epi_df-ness to get the epikeys. I suppose I could have the user specify the grouping columns as arguments instead? But that seems separate from the point you're making that I can assume epi_dfness here and remove this check? Happy to not worry about it, if it's handled.

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+  # Transform each column, using the appropriate lambda column per row.
+  # Note that yj_transform() is vectorized.
+  new_data %<>% left_join(object$lambdas, by = keys)
+  for (col in col_names) {
+    new_data <- new_data %>%
+      rowwise() %>%
```
[the remark this answers is not in the document]
Comment (dshemetov) [reply R/new_epipredict_steps/step_yeo_johnson.R:224]: So we need it for now, because yj_transform is vectorized in x but not in lambda. It can likely be modified to be vectorized in both, but (a) it seems fast enough like this (b) can improve it without breaking changes later if needed.

File: test-yeo-johnson.Rmd [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  mutate(cases = log(cases)) %>%
+  ggplot(aes(time_value, cases)) +
+  geom_line(color = "blue") +
+  geom_line(data = out1 %>% mutate(cases = log(cases)),
+            aes(time_value, cases), color = "green") +
+  geom_line(data = out2 %>% mutate(cases = log(cases)),
+            aes(time_value, cases), color = "red") +
+  facet_wrap(~geo_value, scales = "free_y") +
+  theme_minimal() +
+  labs(title = "Yeo-Johnson transformation", x = "Time", y = "Log Cases")
```
[the remark this answers is not in the document]
Comment (dsweber2) [reply test-yeo-johnson.Rmd:84]: also the blip where it goes literally negative in CA in early July is a bit concerning

[the remark this answers is not in the document]
Comment (dsweber2) [reply test-yeo-johnson.Rmd:84]: looked into it, it's an actual negative value in the raw signal
```
filtered_data %>% filter(geo_value == "ca", time_value == "2021-06-29")
An `epi_df` object, 1 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2024-03-20

# A tibble: 1 × 3
  geo_value time_value cases
  <chr>     <date>     <dbl>
1 ca        2021-06-29 -3940
> out1 %>% filter(geo_value == "ca", time_value == "2021-06-29")
An `epi_df` object, 1 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2024-03-20

# A tibble: 1 × 3
  geo_value time_value   cases
  <chr>     <date>       <dbl>
1 ca        2021-06-29 -11320.
> out2 %>% filter(geo_value == "ca", time_value == "2021-06-29")
An `epi_df` object, 1 x 3 with metadata:
* geo_type  = state
* time_type = day
* as_of     = 2024-03-20

# A tibble: 1 × 3
  geo_value time_value cases
  <chr>     <date>     <dbl>
1 ca        2021-06-29   NaN
```

File: tests/testthat/test-yeo-johnson.R [hunk 6]
```diff
@@ -0,0 +1,105 @@
+suppressPackageStartupMessages(source(here::here("R", "load_all.R")))
+
+test_that("Yeo-Johnson transformation inverts correctly", {
+  expect_true(
+    map_lgl(seq(-5, 5, 0.1), function(lambda) {
+      map_lgl(seq(0, 10, 0.1), \(x) abs(yj_inverse(yj_transform(x, lambda), lambda) - x) < 0.00001) %>% all()
```
Comment (dsweber2) [inline tests/testthat/test-yeo-johnson.R:6]:
```suggestion
      map_lgl(seq(-10, 10, 0.1), \(x) abs(yj_inverse(yj_transform(x, lambda), lambda) - x) < 0.00001) %>% all()
```
Should be robust to negative values of `x` (it does pass this thankfully). Also want to note that the special lambda values of -1, 0, and 1 are all covered exactly already.

File: R/new_epipredict_steps/layer_yeo_johnson.R [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+    # In this case, we should assume the user wants to transform all outcomes.
+    cli::cli_abort("Not specifying columns to layer Yeo-Johnson is not implemented yet.", call = rlang::caller_env())
+  } else {
+    # In this case, we assume that the user has specified the columns they want
+    # transformed here. We then need to determine the lambda columns for each of
+    # these columns. That is, we need to convert a vector of column names like
+    # c(".pred_ahead_1_case_rate", ".pred_ahead_7_case_rate") to
+    # c("lambda_ahead_1_case_rate", "lambda_ahead_7_case_rate").
+    original_outcome_cols <- str_match(col_names, ".pred_ahead_\\d+_(.*)")[, 2]
+    if (all(original_outcome_cols %nin% names(components$mold$outcomes))) {
```
Comment (dsweber2) [inline R/new_epipredict_steps/layer_yeo_johnson.R:158]:
```suggestion
    original_outcome_cols <- str_match(col_names, ".pred_(ahead_\\d+_.*)")[, 2]
    if (all(original_outcome_cols %nin% names(components$mold$outcomes))) {
```
unsure about whether you actually want `ahead_7_cases` generally, but this is what matches `names(components$mold$outcomes)` for the test you're currently skipping

Comment (dsweber2) [reply R/new_epipredict_steps/layer_yeo_johnson.R:158]: turns out, the main problem is actually that the layer doesn't undo the scaling for death_rate. if you also include that, this works fine

File: tests/testthat/test-yeo-johnson.R [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+  expect_true(".lambda_case_rate" %in% names(tr$steps[[1]]$lambdas))
+  expect_true(".lambda_death_rate" %in% names(tr$steps[[1]]$lambdas))
+  expect_true(is.numeric(tr$steps[[1]]$lambdas$.lambda_case_rate))
+  expect_true(is.numeric(tr$steps[[1]]$lambdas$.lambda_death_rate))
+
+  # TODO: Make sure that the inverse transformation works
+  skip("TODO")
+  f <- frosting() %>%
+    layer_predict() %>%
+    layer_epi_YeoJohnson(.pred_ahead_0_case_rate)
```
Comment (dsweber2) [inline tests/testthat/test-yeo-johnson.R:68]:
```suggestion
    layer_epi_YeoJohnson(.pred_ahead_0_case_rate, .pred_ahead_0_death_rate)
```

should undo both the steps done above

File: tests/testthat/test-yeo-johnson.R [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  wf <- epi_workflow(r, linear_reg()) %>%
+    fit(filtered_data) %>%
+    add_frosting(f)
+  out1 <- filtered_data %>% as_tibble() %>% slice_max(time_value, by = geo_value)
+  # debugonce(slather.layer_epi_YeoJohnson)
+  out2 <- forecast(wf) %>% rename(case_rate = .pred)
```
Comment (dsweber2) [inline tests/testthat/test-yeo-johnson.R:74]:
```suggestion
  out2 <- forecast(wf) %>% rename(case_rate = .pred_ahead_0_case_rate, death_rate = .pred_ahead_0_death_rate)
```
a bit unwieldy, not sure there's really a better option

File: test-yeo-johnson.Rmd (hunk 5 above)
[the remark this answers is not in the document]
Comment (dshemetov) [reply test-yeo-johnson.Rmd]: Yea it's a huge data anomaly. I wonder if we should fix this test dataset. On the one hand, it's educational of data reality, on the other hand, what the hell.

[the remark this answers is not in the document]
Comment (dsweber2) [reply test-yeo-johnson.Rmd]: Yeah I take as a demo that it will do ~ the right thing in the presence of both positive and negative values, just wanted to check it was legit and not an artifact of implementation problems.

The difference in scaling for NY & CA is a bit confusing. I forget exactly what the optimization routine is minimizing in its choice of lambda; iirc it's literally the variance? Probably worth noting in the description of the step

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 10]
```diff
@@ -0,0 +1,399 @@
...
+  mu_t <- mean(y_t)
+  var_t <- var(y_t) * (n - 1) / n
+  res <- -.5 * n * log(var_t) + (lambda - 1) * const
+  res
+}
+
+## eliminates missing data and returns -llh
+yj_obj <- function(lam, dat, ind_neg, const) {
+  ll_yj(lambda = lam, y = dat, ind_neg = ind_neg, const = const)
+}
```
Comment (dsweber2) [inline R/new_epipredict_steps/step_yeo_johnson.R:333]: seems to be entirely redundant as a function?

File: R/new_epipredict_steps/step_yeo_johnson.R [hunk 11]
```diff
@@ -0,0 +1,399 @@
...
+
+
+## Helper for the log-likelihood calc for eq 3.1 of Yeo, I. K.,
+## & Johnson, R. A. (2000). A new family of power transformations
+## to improve normality or symmetry. Biometrika. page 957
+ll_yj <- function(lambda, y, ind_neg, const, eps = 0.001) {
+  n <- length(y)
+  y_t <- yj_transform(y, lambda, ind_neg)
+  mu_t <- mean(y_t)
```
Comment (dsweber2) [inline R/new_epipredict_steps/step_yeo_johnson.R:324]: seems like it's inherited, but this is completely unused??

File: R/new_epipredict_steps/step_yeo_johnson.R (hunk 10 above)
Comment (dshemetov) [reply R/new_epipredict_steps/step_yeo_johnson.R:333]: but yes, it's just a wrapper from the original code

File: R/new_epipredict_steps/step_yeo_johnson.R (hunk 11 above)
Comment (dshemetov) [reply R/new_epipredict_steps/step_yeo_johnson.R:324]: yup 🤷

Comment (dsweber2) [reply R/new_epipredict_steps/step_yeo_johnson.R:324]: ok, so 3.1 involves λ, μ, and v (well, σ^2, but that's just the (v)ariance), but if you substitute things from 3.2 after optimizing over μ and v, the term with μ drops out as a constant, leaving us with 326 (they've already left out n/2*log(2π))

**Comment by dshemetov** [comment]
Replaced with https://github.com/cmu-delphi/epipredict/pull/451
